File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/screens/UserDetailsScreen.kt
package com.usesmileid.sampleapps.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Row
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.smileid.designsystem.smileCardStrokeWidth
import com.usesmileid.sampleapps.ui.UseSmileIDSampleTestIds
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleButton
import com.usesmileid.sampleapps.ui.components.ChevronDownGlyph
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleKeyValueEditRow
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleProfileRow
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleSectionSurface
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleSettingRowDivider
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleSwitch
import com.usesmileid.sampleapps.ui.components.UseSmileIDSampleTopAppBar
import androidx.compose.ui.graphics.Color
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleProfile
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleProfiles
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleUserDetails
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleUserDetailsRequirement
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleUserField
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** The Consent Details Form, shown for every product. Sample-owned and ahead of the flow, which is why it needs no SDK change. */
@Composable
fun UserDetailsScreen(
    productLabel: String,
    details: UseSmileIDSampleUserDetails,
    /** Who this run is for; null while there is no profile, when the form offers to create one. */
    profile: UseSmileIDSampleProfile?,
    profileColor: Color,
    saveToProfile: Boolean,
    onFieldChange: (UseSmileIDSampleUserField, String) -> Unit,
    onSaveToProfileChange: (Boolean) -> Unit,
    onProfileClick: () -> Unit,
    onBack: () -> Unit,
    onContinue: () -> Unit,
    modifier: Modifier = Modifier,
    contentPadding: PaddingValues = PaddingValues(),
    /** What is still outstanding once the token's own bindings are taken off the SDK's rule. */
    requirement: UseSmileIDSampleUserDetailsRequirement = UseSmileIDSampleUserDetailsRequirement(),
    organisation: String = "",
    onOrganisationChange: (String) -> Unit = {},
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .testTag(UseSmileIDSampleTestIds.USER_DETAILS_SCREEN),
    ) {
        UseSmileIDSampleTopAppBar(title = productLabel, onBack = onBack)
        LazyColumn(
            modifier = Modifier.weight(1f),
            contentPadding = contentPadding,
            verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
        ) {
            item {
                UseSmileIDSampleProfileRow(
                    organisation = profile?.title ?: UseSmileIDSampleProfiles.NO_PROFILE_LABEL,
                    supportingText = if (profile == null) "Your details below will create one" else "Tap to switch profile",
                    initials = profile?.initials.orEmpty(),
                    avatarColor = profileColor,
                    selected = false,
                    onClick = onProfileClick,
                    modifier = Modifier.padding(horizontal = SmileDimens.spacingMd),
                    testId = UseSmileIDSampleTestIds.USER_DETAILS_PROFILE,
                    trailing = { ChevronDownGlyph(tint = UseSmileIDSampleTheme.colors.textMuted) },
                )
            }
            item {
                UseSmileIDSampleSectionSurface(
                    modifier = Modifier.padding(horizontal = SmileDimens.spacingMd),
                    label = "YOUR DETAILS",
                ) {
                    if (profile == null) {
                        UseSmileIDSampleKeyValueEditRow(
                            label = "Profile name (optional)",
                            value = organisation,
                            onValueChange = onOrganisationChange,
                            placeholder = "Shown on the consent screen",
                            required = false,
                            testId = UseSmileIDSampleTestIds.userDetailsField(ORGANISATION_FIELD_ID),
                        )
                        UseSmileIDSampleSettingRowDivider()
                    }
                    UseSmileIDSampleUserField.entries.forEachIndexed { index, field ->
                        if (index > 0) UseSmileIDSampleSettingRowDivider()
                        // Shown as provided, not asked again — the value is vaulted, so it
                        // cannot be prefilled either.
                        val supplied = requirement.supplies(field)
                        UseSmileIDSampleKeyValueEditRow(
                            label = requirement.labelFor(field),
                            value = if (supplied) "" else field.read(details),
                            onValueChange = { onFieldChange(field, it) },
                            placeholder = if (supplied) "Provided by token" else field.placeholder,
                            required = false,
                            enabled = !supplied,
                            testId = UseSmileIDSampleTestIds.userDetailsField(field.id),
                        )
                    }
                }
            }
            item {
                Text(
                    text = if (details.satisfies(requirement)) "Tap any field to edit." else requirement.prompt,
                    style = UseSmileIDSampleTheme.type.textStyleCaption,
                    color = UseSmileIDSampleTheme.colors.textMuted,
                    modifier = Modifier
                        .testTag(UseSmileIDSampleTestIds.USER_DETAILS_HINT)
                        .padding(horizontal = SmileDimens.spacingMd),
                )
            }
            if (details.satisfies(requirement) && (profile == null || details != profile.defaults)) {
                item {
                    Surface(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(horizontal = SmileDimens.spacingMd),
                        shape = RoundedCornerShape(SmileDimens.radiusSurface),
                        color = UseSmileIDSampleTheme.colors.surface,
                        border = BorderStroke(smileCardStrokeWidth, UseSmileIDSampleTheme.colors.cardStroke),
                    ) {
                        // One line of body text beside the switch: no icon and no supporting line, so not a SettingRow.
                        Row(
                            modifier = Modifier.padding(
                                start = SmileDimens.spacingMd,
                                end = SmileDimens.spacingSm,
                                top = SmileDimens.spacingSm,
                                bottom = SmileDimens.spacingSm,
                            ),
                            horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(
                                text = profile?.let { "Save to ${it.title}" } ?: "Save as a new profile",
                                style = UseSmileIDSampleTheme.type.textStyleSubtitle.copy(fontSize = REMEMBER_TEXT_SIZE),
                                color = UseSmileIDSampleTheme.colors.textBody,
                                modifier = Modifier.weight(1f),
                            )
                            UseSmileIDSampleSwitch(
                                checked = saveToProfile,
                                onCheckedChange = onSaveToProfileChange,
                                testId = UseSmileIDSampleTestIds.REMEMBER_DETAILS_SWITCH,
                            )
                        }
                    }
                }
            }
            item { Spacer(modifier = Modifier.height(SmileDimens.spacingLg)) }
        }
        UseSmileIDSampleButton(
            text = "Continue",
            onClick = onContinue,
            enabled = details.satisfies(requirement),
            modifier = Modifier.padding(SmileDimens.spacingMd),
            testId = UseSmileIDSampleTestIds.USER_DETAILS_CONTINUE,
        )
    }
}

private val REMEMBER_TEXT_SIZE = 13.5.sp

/** The organisation row's id suffix, beside the four user fields'. */
internal const val ORGANISATION_FIELD_ID = "organisation"

File: ios/SampleUI/Sources/SampleUI/Screens/UserDetailsScreen.swift
import SwiftUI

/// What the Consent Details Form draws, plus what the token has already bound.
public struct UseSmileIDSampleUserDetailsState: Equatable {
  public var productLabel: String
  public var details: UseSmileIDSampleUserDetails
  /// Who this run is for; nil while there is no profile, when the form offers to create one.
  public var profile: UseSmileIDSampleProfile?
  public var profileIndex: Int
  public var saveToProfile: Bool
  /// The new profile's name, asked only while there is no profile.
  public var organisation: String
  public var requirement: UseSmileIDSampleUserDetailsRequirement

  public init(
    productLabel: String,
    details: UseSmileIDSampleUserDetails = UseSmileIDSampleUserDetails(),
    profile: UseSmileIDSampleProfile? = nil,
    profileIndex: Int = 0,
    saveToProfile: Bool = true,
    organisation: String = "",
    requirement: UseSmileIDSampleUserDetailsRequirement = UseSmileIDSampleUserDetailsRequirement()
  ) {
    self.productLabel = productLabel
    self.details = details
    self.profile = profile
    self.profileIndex = profileIndex
    self.saveToProfile = saveToProfile
    self.organisation = organisation
    self.requirement = requirement
  }

  var isSatisfied: Bool {
    details.satisfies(requirement)
  }

  /// Only once there is something to keep: valid details that no profile holds yet.
  var offersSave: Bool {
    isSatisfied && details != profile?.defaults
  }

  var saveLabel: String {
    profile.map { "Save to \($0.title)" } ?? "Save as a new profile"
  }
}

/// The Consent Details Form, shown for every product before the SDK flow starts.
public struct UserDetailsScreen: View {
  private let state: UseSmileIDSampleUserDetailsState
  private let onFieldChange: (UseSmileIDSampleUserField, String) -> Void
  private let onSaveToProfileChange: (Bool) -> Void
  private let onOrganisationChange: (String) -> Void
  private let onProfileTap: () -> Void
  private let onBack: () -> Void
  private let onContinue: () -> Void

  @Environment(\.useSmileIDSampleColors) private var colors

  public init(
    state: UseSmileIDSampleUserDetailsState,
    onFieldChange: @escaping (UseSmileIDSampleUserField, String) -> Void,
    onSaveToProfileChange: @escaping (Bool) -> Void,
    onOrganisationChange: @escaping (String) -> Void = { _ in },
    onProfileTap: @escaping () -> Void,
    onBack: @escaping () -> Void,
    onContinue: @escaping () -> Void
  ) {
    self.state = state
    self.onFieldChange = onFieldChange
    self.onSaveToProfileChange = onSaveToProfileChange
    self.onOrganisationChange = onOrganisationChange
    self.onProfileTap = onProfileTap
    self.onBack = onBack
    self.onContinue = onContinue
  }

  public var body: some View {
    VStack(spacing: 0) {
      UseSmileIDSampleTopAppBar(title: state.productLabel, onBack: onBack)
      ScrollView {
        VStack(alignment: .leading, spacing: SmileSpacing.spacingXs) {
          profileRow
          fields
          hint
          if state.offersSave {
            rememberCard
          }
        }
        .padding(.horizontal, SmileSpacing.spacingMd)
      }
      .useSmileIDSampleTestId(UseSmileIDSampleTestIds.userDetailsScreen)
      UseSmileIDSampleButton(
        text: "Continue",
        enabled: state.isSatisfied,
        testId: UseSmileIDSampleTestIds.userDetailsContinue,
        action: onContinue
      )
      .padding(SmileSpacing.spacingMd)
    }
    .background(colors.background)
  }

  /// Says whose details these are, and switches in one tap: the whole reason profiles exist.
  private var profileRow: some View {
    UseSmileIDSampleProfileRow(
      organisation: state.profile?.title ?? UseSmileIDSampleProfiles.noProfileLabel,
      supportingText: state.profile == nil ? "Your details below will create one" : "Tap to switch profile",
      initials: state.profile?.initials ?? "",
      selected: false,
      avatarColor: useSmileIDSampleAvatarColor(profileIndex: state.profileIndex),
      testId: UseSmileIDSampleTestIds.userDetailsProfile,
      onTap: onProfileTap
    ) {
      UseSmileIDSampleIcon(SmileIcons.chevronDown, tint: colors.textMuted, size: 12)
    }
  }

  private var fields: some View {
    UseSmileIDSampleSectionSurface(label: "YOUR DETAILS") {
      if state.profile == nil {
        UseSmileIDSampleKeyValueEditRow(
          label: "Organisation (optional)",
          value: Binding(get: { state.organisation }, set: onOrganisationChange),
          placeholder: "Shown on the consent screen",
          testId: UseSmileIDSampleTestIds.userDetailsField(Self.organisationFieldId)
        )
        UseSmileIDSampleRowDivider()
      }
      ForEach(Array(UseSmileIDSampleUserField.allCases.enumerated()), id: \.element) { index, field in
        if index > 0 {
          UseSmileIDSampleRowDivider()
        }
        row(field)
      }
    }
  }

  /// A supplied value is vaulted, so it is shown as provided rather than prefilled.
  private func row(_ field: UseSmileIDSampleUserField) -> some View {
    let supplied = state.requirement.supplies(field)
    return UseSmileIDSampleKeyValueEditRow(
      label: state.requirement.label(for: field),
      value: Binding(
        get: { supplied ? "" : field.read(state.details) },
        set: { onFieldChange(field, $0) }
      ),
      placeholder: supplied ? "Provided by token" : field.placeholder,
      enabled: !supplied,
      keyboardType: field.keyboardType,
      testId: UseSmileIDSampleTestIds.userDetailsField(field.id)
    )
  }

  private var hint: some View {
    UseSmileIDSampleText(
      state.isSatisfied ? "Tap any field to edit." : state.requirement.prompt,
      style: UseSmileIDSampleTheme.type.textStyleCaption
    )
    .foregroundColor(colors.textMuted)
    .useSmileIDSampleTestId(UseSmileIDSampleTestIds.userDetailsHint)
  }

  /// One line of body text beside the switch, so not a `SettingRow`.
  private var rememberCard: some View {
    HStack(spacing: SmileSpacing.spacingSm) {
      UseSmileIDSampleText(
        state.saveLabel,
        style: UseSmileIDSampleTheme.type.textStyleSubtitle.with(size: 13.5)
      )
      .foregroundColor(colors.textBody)
      .frame(maxWidth: .infinity, alignment: .leading)
      UseSmileIDSampleSwitch(
        isOn: Binding(get: { state.saveToProfile }, set: onSaveToProfileChange),
        testId: UseSmileIDSampleTestIds.rememberDetailsSwitch
      )
    }
    .padding(.leading, SmileSpacing.spacingMd)
    .padding(.trailing, SmileSpacing.spacingSm)
    .padding(.vertical, SmileSpacing.spacingSm)
    .frame(maxWidth: .infinity)
    .background(
      RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.card, style: .continuous)
        .fill(colors.surface)
    )
    .overlay(
      RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.card, style: .continuous)
        .strokeBorder(colors.cardStroke, lineWidth: smileCardStrokeWidth)
    )
  }
}

extension UserDetailsScreen {
  /// The organisation row's id suffix, beside the four user fields'.
  static let organisationFieldId = "organisation"
}

extension UseSmileIDSampleUserField {
  var keyboardType: UIKeyboardType {
    switch self {
    case .email: .emailAddress
    case .phone: .phonePad
    default: .default
    }
  }
}
